File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic/BoundLocal.cs
using System.Diagnostics;
using Microsoft.CodeAnalysis.VisualBasic.Symbols;

namespace Microsoft.CodeAnalysis.VisualBasic
{
	internal sealed class BoundLocal : BoundExpression
	{
		private readonly LocalSymbol _LocalSymbol;

		private readonly bool _IsLValue;

		public override Symbol ExpressionSymbol => LocalSymbol;

		public override ConstantValue ConstantValueOpt
		{
			get
			{
				if (base.HasErrors || TypeSymbolExtensions.IsErrorType(base.Type))
				{
					return null;
				}
				return LocalSymbol.GetConstantValue(null);
			}
		}

		public LocalSymbol LocalSymbol => _LocalSymbol;

		public override bool IsLValue => _IsLValue;

		public BoundLocal(SyntaxNode syntax, LocalSymbol localSymbol, TypeSymbol type, bool hasErrors)
			: this(syntax, localSymbol, !localSymbol.IsReadOnly, type, hasErrors)
		{
		}

		public BoundLocal(SyntaxNode syntax, LocalSymbol localSymbol, TypeSymbol type)
			: this(syntax, localSymbol, !localSymbol.IsReadOnly, type)
		{
		}

		protected override BoundExpression MakeRValueImpl()
		{
			return MakeRValue();
		}

		public new BoundLocal MakeRValue()
		{
			if (_IsLValue)
			{
				return Update(_LocalSymbol, isLValue: false, base.Type);
			}
			return this;
		}

		public BoundLocal(SyntaxNode syntax, LocalSymbol localSymbol, bool isLValue, TypeSymbol type, bool hasErrors)
			: base(BoundKind.Local, syntax, type, hasErrors)
		{
			_LocalSymbol = localSymbol;
			_IsLValue = isLValue;
		}

		public BoundLocal(SyntaxNode syntax, LocalSymbol localSymbol, bool isLValue, TypeSymbol type)
			: base(BoundKind.Local, syntax, type)
		{
			_LocalSymbol = localSymbol;
			_IsLValue = isLValue;
		}

		[DebuggerStepThrough]
		public override BoundNode Accept(BoundTreeVisitor visitor)
		{
			return visitor.VisitLocal(this);
		}

		public BoundLocal Update(LocalSymbol localSymbol, bool isLValue, TypeSymbol type)
		{
			if ((object)localSymbol != LocalSymbol || isLValue != IsLValue || (object)type != base.Type)
			{
				BoundLocal boundLocal = new BoundLocal(base.Syntax, localSymbol, isLValue, type, base.HasErrors);
				boundLocal.CopyAttributes(this);
				return boundLocal;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/BoundLocal.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
Imports System
Imports System.Diagnostics

Namespace Microsoft.CodeAnalysis.VisualBasic
	Friend NotInheritable Class BoundLocal
		Inherits BoundExpression
		Private ReadOnly _LocalSymbol As Microsoft.CodeAnalysis.VisualBasic.Symbols.LocalSymbol

		Private ReadOnly _IsLValue As Boolean

		Public Overrides ReadOnly Property ConstantValueOpt As ConstantValue
			Get
				Dim constantValue As Microsoft.CodeAnalysis.ConstantValue
				If (MyBase.HasErrors OrElse MyBase.Type.IsErrorType()) Then
					constantValue = Nothing
				Else
					constantValue = Me.LocalSymbol.GetConstantValue(Nothing)
				End If
				Return constantValue
			End Get
		End Property

		Public Overrides ReadOnly Property ExpressionSymbol As Symbol
			Get
				Return Me.LocalSymbol
			End Get
		End Property

		Public Overrides ReadOnly Property IsLValue As Boolean
			Get
				Return Me._IsLValue
			End Get
		End Property

		Public ReadOnly Property LocalSymbol As Microsoft.CodeAnalysis.VisualBasic.Symbols.LocalSymbol
			Get
				Return Me._LocalSymbol
			End Get
		End Property

		Public Sub New(ByVal syntax As SyntaxNode, ByVal localSymbol As Microsoft.CodeAnalysis.VisualBasic.Symbols.LocalSymbol, ByVal type As TypeSymbol, ByVal hasErrors As Boolean)
			MyClass.New(syntax, localSymbol, Not localSymbol.IsReadOnly, type, hasErrors)
		End Sub

		Public Sub New(ByVal syntax As SyntaxNode, ByVal localSymbol As Microsoft.CodeAnalysis.VisualBasic.Symbols.LocalSymbol, ByVal type As TypeSymbol)
			MyClass.New(syntax, localSymbol, Not localSymbol.IsReadOnly, type)
		End Sub

		Public Sub New(ByVal syntax As SyntaxNode, ByVal localSymbol As Microsoft.CodeAnalysis.VisualBasic.Symbols.LocalSymbol, ByVal isLValue As Boolean, ByVal type As TypeSymbol, ByVal hasErrors As Boolean)
			MyBase.New(BoundKind.Local, syntax, type, hasErrors)
			Me._LocalSymbol = localSymbol
			Me._IsLValue = isLValue
		End Sub

		Public Sub New(ByVal syntax As SyntaxNode, ByVal localSymbol As Microsoft.CodeAnalysis.VisualBasic.Symbols.LocalSymbol, ByVal isLValue As Boolean, ByVal type As TypeSymbol)
			MyBase.New(BoundKind.Local, syntax, type)
			Me._LocalSymbol = localSymbol
			Me._IsLValue = isLValue
		End Sub

		<DebuggerStepThrough>
		Public Overrides Function Accept(ByVal visitor As BoundTreeVisitor) As BoundNode
			Return visitor.VisitLocal(Me)
		End Function

		Public Shadows Function MakeRValue() As Microsoft.CodeAnalysis.VisualBasic.BoundLocal
			Dim boundLocal As Microsoft.CodeAnalysis.VisualBasic.BoundLocal
			boundLocal = If(Not Me._IsLValue, Me, Me.Update(Me._LocalSymbol, False, MyBase.Type))
			Return boundLocal
		End Function

		Protected Overrides Function MakeRValueImpl() As BoundExpression
			Return Me.MakeRValue()
		End Function

		Public Function Update(ByVal localSymbol As Microsoft.CodeAnalysis.VisualBasic.Symbols.LocalSymbol, ByVal isLValue As Boolean, ByVal type As TypeSymbol) As Microsoft.CodeAnalysis.VisualBasic.BoundLocal
			Dim boundLocal As Microsoft.CodeAnalysis.VisualBasic.BoundLocal
			If (CObj(localSymbol) <> CObj(Me.LocalSymbol) OrElse isLValue <> Me.IsLValue OrElse CObj(type) <> CObj(MyBase.Type)) Then
				Dim boundLocal1 As Microsoft.CodeAnalysis.VisualBasic.BoundLocal = New Microsoft.CodeAnalysis.VisualBasic.BoundLocal(MyBase.Syntax, localSymbol, isLValue, type, MyBase.HasErrors)
				boundLocal1.CopyAttributes(Me)
				boundLocal = boundLocal1
			Else
				boundLocal = Me
			End If
			Return boundLocal
		End Function
	End Class
End Namespace